lunatrace/bsl/ingest-worker/pkg/epss/fetch.go

```diff
@@ -0,0 +1,120 @@
+// Copyright by LunaSec (owned by Refinery Labs, Inc)
+//
+// Licensed under the Business Source License v1.1
+// (the "License"); you may not use this file except in compliance with the
+// License. You may obtain a copy of the License at
+//
+// https://github.com/lunasec-io/lunasec/blob/master/licenses/BSL-LunaTrace.txt
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package epss
+
+import (
+  "bytes"
+  "compress/gzip"
+  "encoding/csv"
+  "fmt"
+  "github.com/rs/zerolog/log"
+  "io"
+  "net/http"
+  "strconv"
+  "time"
+)
+
+// EpssVulnerability has a CVE, Epss, and Percentile
+type EpssVulnerability struct {
+  Cve        string
+  Epss       float64
+  Percentile float64
+}
+
+func FetchEpssScores() (epssScores []EpssVulnerability, err error) {
+  // Get the current date and time
+  now := time.Now()
+
+  // Format the date and time according to the desired format
+  dateStr := now.Format("2006-01-02")
+
+  // Use the formatted date string to construct the URL
+  url := fmt.Sprintf("https://epss.cyentia.com/epss_scores-%s.csv.gz", dateStr)
+
+  // Download the file
+  response, err := http.Get(url)
+  if err != nil {
+    fmt.Printf("Error downloading file: %v\n", err)
+    return
+  }
+  defer response.Body.Close()
+
+  body, err := io.ReadAll(response.Body)
+  if err != nil {
+    log.Error().Err(err).Msg("failed to read cwe data")
+    return
+  }
+
+  buf := bytes.NewReader(body)
+
+  // Create a new gzip reader
+  gzipReader, err := gzip.NewReader(buf)
+  if err != nil {
+    fmt.Printf("Error creating gzip reader: %v\n", err)
+    return
+  }
+  defer gzipReader.Close()
+
+  // Create a new CSV reader
+  csvReader := csv.NewReader(gzipReader)
+
+  // We have to set this or else we get a "wrong number of fields" error for some reason
```

Comment:
```
	// FieldsPerRecord is the number of expected fields per record.
	// If FieldsPerRecord is positive, Read requires each record to
	// have the given number of fields. If FieldsPerRecord is 0, Read sets it to
	// the number of fields in the first record, so that future records must
	// have the same field count. If FieldsPerRecord is negative, no check is
	// made and records may have a variable number of fields.
```
[Godoc](https://pkg.go.dev/encoding/csv#:~:text=number%20of%20fields.-,FieldsPerRecord,-int%0A%0A%09//%20If%20LazyQuotes) can usually help when you are seeing weird semantic behavior of a package. The error comes from there being a varying number of fields in the first few rows of the file.

Reply:
It's because the first line of the CVS is a weird format. It looks like I just needed to set `csvReader.Comment = '#'` to fix that. Thanks :pray: